File: doc/codes/python/examples/tutorial/moving-focus-stm.py
import asyncio

import numpy as np

import autd3_link_echocat as echocat
from autd3 import Client, ClientConfig
from autd3.commands import FociStm, FociStmOption
from autd3.geometry import Autd3, Geometry
from autd3.units import Hz
from autd3.value import ControlPoint, ControlPoints

async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    async with await Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        ClientConfig(),
    ) as client:
        center = geometry.center() + np.array([0.0, 0.0, 150.0])

        # ANCHOR: stm
        points = [
            ControlPoints([ControlPoint(center + np.array([20.0, 0.0, 0.0]))]),
            ControlPoints([ControlPoint(center + np.array([-20.0, 0.0, 0.0]))]),
        ]
        builder = client.datagram_builder()
        builder.push(FociStm(0.5 * Hz, points, FociStmOption()))
        for frame in builder.build():
            await client.send_checked(frame)
        # ANCHOR_END: stm


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/moving-focus-stm.rs
use anyhow::Result;

use autd3_rs::commands::{FociStm, FociStmOption, Modulation, SetSilencer};
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{Hz, mm};
use autd3_rs::value::SamplingConfig;
use autd3_rs::{Client, ClientConfig, ControlPoints};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);

    // ANCHOR: stm
    let points = [
        ControlPoints::from(center + offset(20.0 * mm, 0.0 * mm, 0.0 * mm)),
        ControlPoints::from(center + offset(-20.0 * mm, 0.0 * mm, 0.0 * mm)),
    ];
    let mut builder = client.datagram_builder();
    builder.push(FociStm::new(0.5 * Hz, &points, FociStmOption::default()));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }
    // ANCHOR_END: stm

    client.close().await?;
    Ok(())
}
